PC/launcher2.c

```diff
@@ -871,6 +871,77 @@ _findCommand(SearchInfo *search, const wchar_t *command, int commandLength)
 }
 
 
+int
+_useShebangAsExecutable(SearchInfo *search, const wchar_t *shebang, int shebangLength)
+{
+    wchar_t buffer[MAXLEN];
+    wchar_t script[MAXLEN];
+    wchar_t command[MAXLEN];
+
+    int commandLength = 0;
+    int inQuote = 0;
+    int afterSlash = 0;
+
+    if (!shebang || !shebangLength) {
+        return 0;
+    }
+
+    wchar_t *pC = command;
+    for (int i = 0; i < shebangLength; ++i) {
+        wchar_t c = shebang[i];
+        if ((isspace(c) || c == L'\r' || c == L'\n') && !inQuote) {
+            commandLength = i;
+            break;
+        } else if (c == L'"') {
+            if (!afterSlash) {
+                // non-escaped quote. either way, we don't add it to the path
+                inQuote = !inQuote;
+            }
+            afterSlash = 0;
+        } else if (c == L'/' || c == L'\\') {
+            if (afterSlash) {
+                // escaped slash
+                *pC++ = L'\\';
+                afterSlash = 0;
+            } else {
+                afterSlash = 1;
+            }
+        } else {
+            if (afterSlash) {
+                *pC++ = L'\\';
+                afterSlash = 0;
+            }
+            *pC++ = c;
+        }
+    }
+    *pC = L'\0';
+
+    if (!GetCurrentDirectoryW(MAXLEN, buffer) ||
+        wcsncpy_s(script, MAXLEN, search->scriptFile, search->scriptFileLength) ||
+        FAILED(PathCchCombineEx(buffer, MAXLEN, buffer, script,
+                                PATHCCH_ALLOW_LONG_PATHS)) ||
+        FAILED(PathCchRemoveFileSpec(buffer, MAXLEN)) ||
+        FAILED(PathCchCombineEx(buffer, MAXLEN, buffer, command,
+                                PATHCCH_ALLOW_LONG_PATHS))
+    ) {
+        return RC_NO_MEMORY;
+    }
```

Comment:
The filepath in a shebang gets resolved against the current working directory, e.g. `GetFullPathNameW()`, not the script directory. That's just how shebangs work in Unix. 

For example, my local bin directory in Ubuntu contains a "python" symlink. Beside it, I created a test script with the shebang "#!python". The system doesn't look for "python" beside the script that contains the shebang. Executing the script only works if the current directory has a "python" binary:

```pycon
>>> try: subprocess.call(['test_script'])
... except OSError as e: print(e)
... 
[Errno 13] Permission denied: 'test_script'

>>> cwd = os.path.dirname(shutil.which('test_script'))
>>> subprocess.call(['test_script'], cwd=cwd)
Hello, World!
0
```

The launcher inherits the shebang concept from Unix, and it should use the same behavior.

Reply:
I _do_ need to restore the `#! python` behaviour though. Previously that was a specific shebang template, and so once that's back then `#!pythonx.y` (for any `x.y` including nothing) will be treated as if it were passed as the tag.